akka-cluster/src/multi-jvm/scala/akka/cluster/StreamRefSpec.scala

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+import akka.Done
+import akka.actor.Actor
+import akka.actor.ActorIdentity
+import akka.actor.ActorRef
+import akka.actor.Identify
+import akka.actor.Props
+import akka.pattern.pipe
+import akka.remote.testkit.MultiNodeConfig
+import akka.remote.testkit.MultiNodeSpec
+import akka.remote.transport.ThrottlerTransportAdapter.Direction
+import akka.stream.ActorMaterializer
+import akka.stream.RemoteStreamRefActorTerminatedException
+import akka.stream.SinkRef
+import akka.stream.SourceRef
+import akka.stream.scaladsl.Keep
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Source
+import akka.stream.scaladsl.StreamRefs
+import akka.stream.testkit.TestSubscriber
+import akka.stream.testkit.scaladsl.TestSink
+import akka.testkit._
+import com.typesafe.config.ConfigFactory
+
+object StreamRefSpec extends MultiNodeConfig {
+  val first = role("first")
+  val second = role("second")
+  val third = role("third")
+
+  commonConfig(debugConfig(on = false).
+    withFallback(ConfigFactory.parseString("""
+        akka.cluster {
+          auto-down-unreachable-after = 1s
+        }""")).
+    withFallback(MultiNodeClusterSpec.clusterConfig))
+
+  testTransport(on = true)
+
+  case class RequestLogs(streamId: Int)
+  case class LogsOffer(streamId: Int, sourceRef: SourceRef[String])
+
+  object DataSource {
+    def props(streamLifecycleProbe: ActorRef): Props =
+      Props(new DataSource(streamLifecycleProbe))
+  }
+
+  class DataSource(streamLifecycleProbe: ActorRef) extends Actor {
+    import context.dispatcher
+    implicit val mat = ActorMaterializer()(context)
+
+    def receive = {
+      case RequestLogs(streamId) ⇒
+        // materialize the SourceRef:
+        val (done: Future[Done], ref: Future[SourceRef[String]]) =
+          Source(1 to 1000)
+            .map(n ⇒ s"elem-$n")
+            .watchTermination() {
+              case (_, d) ⇒ d
+            }
+            .toMat(StreamRefs.sourceRef())(Keep.both)
+            .mapMaterializedValue { m ⇒
+              streamLifecycleProbe ! s"started-$streamId"
+              m
+            }
+            .run()
+
+        done.onComplete(_ ⇒ streamLifecycleProbe ! s"ended-$streamId")
```

Comment:
This is from `watchTermination`, which is upstream so it will be completed with success Done due to cancellation.

On the other side it will be error, which I also check for in the test: `RemoteStreamRefActorTerminatedException`

Reply:
also note that the test will not complete the 1000 elements because that is far less than the `buffer-capacity = 32` of stream refs, and only 3 are requested from other side.